src/matrices.jl

```diff
@@ -0,0 +1,66 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENCE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    UnitMatrix{T}(d)
+    UnitColumn{T}(d)
+    F{T}(d)
+    G{T}(N)
+    H{T}(d)
+
+Elementary matrices as defined by Aitchison 1986.
+
+## Examples
+
+```julia
+julia> UnitMatrix(5)
+julia> UnitColumn{Int}(2)
+julia> G(4)
+julia> H{Float64}(2)
+```
+
+When the type of the elements isn't explicitly defined, `Float64` is used.
+"""
+
+struct UnitMatrix{T} <: AbstractArray{T, 2}
+    d::Int
+end
+UnitMatrix(d::Int) = UnitMatrix{Float64}(d)
+Base.size(J::UnitMatrix) = (J.d, J.d)
+Base.IndexStyle(::Type{<:UnitMatrix}) = IndexLinear()
+Base.getindex(J::UnitMatrix, i::Int, j::Int) = convert(eltype(J), 1)
+
+
+struct UnitColumn{T} <: AbstractArray{T, 1}
+    d::Int
+end
+UnitColumn(d::Int) = UnitColumn{Float64}(d)
```

Comment:
We should default to Bool because then Julia will use its promote rules with other types when needed. If you specify Float64 you are already forcing double precision and inhibiting some applications (ex: GPU with Float32).

Reply:
Since my knowledge of the Julia's compiler is limited, I'd like to document a few bullet points that influenced my decision to use Float64 as a default:

- I saw that the Identity Matrix was implemented using Bool by default, but my idea was to avoid casting during the operations, which can hinder AVX's performance.

- We need to also keep in mind that the auto casting is non-associative and non-commutative, ie,

```
julia> typeof(Float32(1) * true/( true + true ))
Float32

julia> typeof(Float32(1) * (true/( true + true )))
Float64

julia> typeof( true/( true + true ) * Float32(1))
Float64
```

which means that auto casting at compilation time probably isn't possible for all cases in Julia.

- The default promotions in Julia are to Int64 and to Float64, so for GPU(s) I'd sitck to manual casting.

With those bullet points in mind, I see no problem in switching to Bool and Int whenever possible.

Reply:
> * I saw that the Identity Matrix was implemented using Bool by default, but my idea was to avoid casting during the operations, which can hinder AVX's performance.

For some of these special matrices we will optimize the matrix-vector multiplication in a way that doesn't even use the type stored in the struct. For example, multiplying the matrix `J` by any vector `x` is equivalent to producing a new vector where the entries are `sum(x)`. That is the beauty of Julia. If you read the entire LinearAlgebra example with the identity matrix, you will see that multiplications and many other operations are completely erased at compile time. We will do the same here after the basic struct is ready for tests.